Mathlib/CategoryTheory/MorphismProperty/OverAdjunction.lean

```diff
@@ -0,0 +1,145 @@
+/-
+Copyright (c) 2024 Christian Merten. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Christian Merten
+-/
+import Mathlib.CategoryTheory.MorphismProperty.Comma
+import Mathlib.CategoryTheory.Adjunction.Over
+import Mathlib.CategoryTheory.MorphismProperty.Limits
+
+/-!
+# Adjunction of pushforward and pullback in `P.Over Q X`
+
+A morphism `f : X ⟶ Y` defines two functors:
+
+- `Over.map`: post-composition with `f`
+- `Over.pullback`: base-change along `f`
+
+These are adjoint under suitable assumptions on `P` and `Q`.
+
+-/
+
+namespace CategoryTheory.MorphismProperty
+
+open Limits
+
+variable {T : Type*} [Category T] (P Q : MorphismProperty T) [Q.IsMultiplicative]
+variable {X Y : T} (f : X ⟶ Y)
+
+section Map
+
+variable {P} [P.IsStableUnderComposition] (hPf : P f)
+
+variable {f}
+
+/-- If `P` is stable under composition and `f : X ⟶ Y` satisfies `P`,
+this is the functor `P.Over Q X ⥤ P.Over Q Y` given by composing with `f`. -/
+def Over.map : P.Over Q X ⥤ P.Over Q Y :=
+  Comma.mapRight _ (Discrete.natTrans fun _ ↦ f) <| fun X ↦ P.comp_mem _ _ X.prop hPf
+
+@[simp] lemma Over.map_obj_left (A : P.Over Q X) : ((Over.map Q hPf).obj A).left = A.left := rfl
+
+@[simp] lemma Over.map_obj_hom (A : P.Over Q X) : ((Over.map Q hPf).obj A).hom = A.hom ≫ f := rfl
+
+@[simp] lemma Over.map_map_left {A B : P.Over Q X} (g : A ⟶ B) :
+    ((Over.map Q hPf).map g).left = g.left := rfl
+
+lemma Over.mapComp_eq {X Y Z : T} {f : X ⟶ Y} (hf : P f) {g : Y ⟶ Z} (hg : P g) :
+    map Q (P.comp_mem f g hf hg) = map Q hf ⋙ map Q hg := by
+  fapply Functor.ext
+  · simp [map, Comma.mapRight, CategoryTheory.Comma.mapRight]
+  · intro U V k
+    ext
+    simp
+
+/-- `Over.map` commutes with composition. -/
+@[simps!]
+def Over.mapComp {X Y Z : T} {f : X ⟶ Y} (hf : P f) {g : Y ⟶ Z} (hg : P g) :
+    map Q (P.comp_mem f g hf hg) ≅ map Q hf ⋙ map Q hg :=
+  eqToIso (Over.mapComp_eq Q hf hg)
```

Comment:
Unless you intend to use this when `Q` does not respect isos, I would favour the following (which avoids `eqToIso`):
```suggestion
@[simps! hom_app_left inv_app_left]
def Over.mapComp {X Y Z : T} {f : X ⟶ Y} (hf : P f) {g : Y ⟶ Z} (hg : P g) [Q.RespectsIso] :
    map Q (P.comp_mem f g hf hg) ≅ map Q hf ⋙ map Q hg :=
  NatIso.ofComponents (fun X ↦ Over.isoMk (Iso.refl _))
```